sdk/src/utils/graphs/adjacency-pool-graph.ts

```diff
@@ -0,0 +1,149 @@
+import { AddressUtil } from "@orca-so/common-sdk";
+import { Address } from "@project-serum/anchor";
+import {
+  PoolGraph,
+  PoolGraphUtils,
+  PoolTokenPair,
+  Route,
+  RouteFindOptions,
+  RouteMap
+} from "../public/pool-graph";
+
+export class AdjacencyPoolGraph implements PoolGraph {
+  readonly graph: {
+    [k: string]: PoolGraphEdge[];
+  };
+
+  constructor(pools: PoolTokenPair[]) {
+    this.graph = buildPoolGraph(pools);
+  }
+
+  getRoute(startMint: Address, endMint: Address, options?: RouteFindOptions): Route[] {
+    const [startMintKey, endMintKey] = [
+      AddressUtil.toString(startMint),
+      AddressUtil.toString(endMint),
+    ];
+
+    const walkMap = findWalks(
+      [[startMintKey, endMintKey]],
+      this.graph,
+      options?.intermediateTokens.map((token) => AddressUtil.toString(token))
+    );
+
+    return Object.values(walkMap)
+      .map((walks) => {
+        return walks.map((walk) => {
+          return {
+            startMint: startMintKey,
+            endMint: endMintKey,
+            edges: walk,
+          };
+        });
+      })
+      .flatMap((x) => x);
+  }
+
+  getAllRoutes(tokens: [Address, Address][], options?: RouteFindOptions): RouteMap {
+    const tokenPairs = tokens.map(([startMint, endMint]) => {
+      return [AddressUtil.toString(startMint), AddressUtil.toString(endMint)] as const;
+    });
+    const walkMap = findWalks(
+      tokenPairs,
+      this.graph,
+      options?.intermediateTokens.map((token) => AddressUtil.toString(token))
+    );
+
+    const walkEntries = Object.entries(walkMap).map(([routeId, walks]) => {
+      const [startMint, endMint] = routeId.split("-");
+      const paths = walks.map<Route>((walk) => {
+        return {
+          startMint,
+          endMint,
+          edges: walk,
+        };
+      });
+      return [routeId, paths] as const;
+    });
+
+    return Object.fromEntries(walkEntries);
+  }
+}
+
+type AdjacencyPoolGraphMap = Record<string, Array<PoolGraphEdge>>;
+
+type PoolGraphEdge = {
+  address: string;
+  otherToken: string;
+};
+
+// A record of route-id (tokenA-tokenB) to a list of edges
+type PoolWalks = Record<string, string[][]>;
+
+function buildPoolGraph(pools: PoolTokenPair[]) {
+  const poolGraphSet = pools.reduce((poolGraph: Record<string, Set<PoolGraphEdge>>, pool) => {
+    const { address, tokenMintA, tokenMintB } = pool;
+    const [addr, mintA, mintB] = AddressUtil.toPubKeys([address, tokenMintA, tokenMintB]).map(
+      (pk) => pk.toBase58()
+    );
+
+    if (poolGraph[mintA] === undefined) {
+      poolGraph[mintA] = new Set();
+    }
+
+    if (poolGraph[mintB] === undefined) {
+      poolGraph[mintB] = new Set();
+    }
+
+    poolGraph[mintA].add({ address: addr, otherToken: mintB });
+    poolGraph[mintB].add({ address: addr, otherToken: mintA });
+    return poolGraph;
+  }, {});
```

Comment:
set cannot deduplicate for objects.
```suggestion
  const processed = new Set<string>();
  const poolGraphSet = pools.reduce((poolGraph: Record<string, PoolGraphEdge[]>, pool) => {
    const { address, tokenMintA, tokenMintB } = pool;
    const [addr, mintA, mintB] = AddressUtil.toPubKeys([address, tokenMintA, tokenMintB]).map(
      (pk) => pk.toBase58()
    );

    if (processed.has(addr)) {
      return poolGraph;
    }

    if (poolGraph[mintA] === undefined) {
      poolGraph[mintA] = [];
    }

    if (poolGraph[mintB] === undefined) {
      poolGraph[mintB] = [];
    }

    poolGraph[mintA].push({ address: addr, otherToken: mintB });
    poolGraph[mintB].push({ address: addr, otherToken: mintA });
    processed.add(addr);
    return poolGraph;
  }, {});
```